source/source_test.go

```diff
@@ -16,15 +16,617 @@ package source
+
+	ctx := context.Background()
+	testBucket := cfg[config.ConfigKeyGCSBucket]
+	source := &Source{}
+	err := source.Configure(ctx, cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = source.Open(ctx, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testFiles := addObjectsToTheBucket(ctx, t, testBucket, client, 3)
+
+	// read and assert
+	for _, file := range testFiles {
+		_, err := readAndAssert(ctx, t, source, file)
+		if err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+	}
+
+	// make sure the insert action has a different lastModifiedDate
+	// because CDC iterator detects files from after maxLastModifiedDate by initial load
+	time.Sleep(time.Second)
+
+	content := uuid.NewString()
+	testFileName := "test-file"
+	// insert a file to the bucket
+	wc := client.Bucket(testBucket).Object(testFileName).NewWriter(ctx)
+	fmt.Fprint(wc, content)
+	wc.Close()
+
+	obj, err := readWithTimeout(ctx, source, time.Second*15)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	lastReadPosition := obj.Position
+	// the insert should have been detected
+	if strings.Compare(string(obj.Key.Bytes()), testFileName) != 0 {
+		t.Fatalf("expected key: %s, got: %s", testFileName, string(obj.Key.Bytes()))
+	}
+
+	// call teardown to stop iterator and close the client
+	_ = source.Teardown(ctx)
+
+	// start the source process again
+	source1 := &Source{}
+	defer func() {
+		_ = source1.Teardown(ctx)
+	}()
+
+	err = source1.Configure(ctx, cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = source1.Open(ctx, lastReadPosition)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// make sure the insert action has a different lastModifiedDate
+	// because CDC iterator detects files from after maxLastModifiedDate by initial load
+	time.Sleep(time.Second)
+
+	content = uuid.NewString()
+	testFileName = "test-file1"
+	// insert a file to the bucket
+	wc = client.Bucket(testBucket).Object(testFileName).NewWriter(ctx)
+	fmt.Fprint(wc, content)
+	wc.Close()
+
+	obj, err = readWithTimeout(ctx, source1, time.Second*15)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// the insert should have been detected
+	if strings.Compare(string(obj.Key.Bytes()), testFileName) != 0 {
+		t.Fatalf("expected key: %s, got: %s", testFileName, string(obj.Key.Bytes()))
+	}
+}
+
+func TestSource_CDCPosition(t *testing.T) {
+	client, cfg := prepareIntegrationTest(t)
+
+	ctx := context.Background()
+	testBucket := cfg[config.ConfigKeyGCSBucket]
+	source := &Source{}
+	defer func() {
+		_ = source.Teardown(ctx)
+	}()
+
+	err := source.Configure(ctx, cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	addObjectsToTheBucket(ctx, t, testBucket, client, 2)
+
+	testFileName := "file0001" // already exists in the bucket
+	expectedAction := "delete"
+	// Delete a file that exists in the bucket
+	err = client.Bucket(testBucket).Object(testFileName).Delete(context.Background())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	p, err := json.Marshal(position.Position{
+		Key:       "file0001",
+		Timestamp: time.UnixMilli(1634049397),
```

Comment:
This appears to be a magic value.

Reply:
It was a Past time, now i have changed testcase start time.

Reply:
:+1: